src/hotspot/share/classfile/moduleEntry.cpp

```diff
@@ -496,12 +544,17 @@ void ModuleEntry::restore_archived_oops(ClassLoaderData* loader_data) {
   // because it may be affected by archive relocation.
   java_lang_Module::set_module_entry(module_handle(), this);
 
-  if (loader_data->class_loader() != NULL) {
-    java_lang_Module::set_loader(module_handle(), loader_data->class_loader());
+  assert(java_lang_Module::loader(module_handle()) == loader_data->class_loader(),
+         "must be set in dump time");
+
+  if (log_is_enabled(Info, cds, module)) {
+    ResourceMark rm;
+    log_info(cds, module)("Restored archived %s", debug_info());
```

Comment:
I don't know if stringStream needs a resource mark so if it doesn't, you can just have one log line for these.

Reply:
I used `StringStream::as_string(bool c_heap = false)` to return from `debug_info()`, so the result is resource allocated.

Anyway, I can see that this style may be confusing. I'll rewrite it to follow the `print_on(st)` style that's more common where the semantics are well understood. This will make the logging a little more tedious but I think it's OK.